neo/Consensus/ConsensusService.cs

```diff
@@ -99,27 +112,30 @@ private void ChangeTimer(TimeSpan delay)
 
         private void CheckCommits()
         {
-            if (context.CommitPayloads.Count(p => p != null) >= context.M() && context.TransactionHashes.All(p => context.Transactions.ContainsKey(p)))
-            {
-                Block block = context.CreateBlock();
-                Log($"relay block: {block.Hash}");
-                localNode.Tell(new LocalNode.Relay { Inventory = block });
-            }
+            if (context.CommitPayloads.Count(p => p != null) < context.M() || !context.TransactionHashes.All(p => context.Transactions.ContainsKey(p))) return;
+            Block block = context.CreateBlock();
+            Log($"relay block: {block.Hash}");
+            localNode.Tell(new LocalNode.Relay { Inventory = block });
         }
 
         private void CheckExpectedView(byte viewNumber)
         {
             if (context.ViewNumber == viewNumber) return;
-            if (context.ChangeViewPayloads.Count(p => p != null && p.GetDeserializedMessage<ChangeView>().NewViewNumber == viewNumber) >= context.M())
+            if (context.ChangeViewPayloads.Count(p => p != null && p.GetDeserializedMessage<ChangeView>().NewViewNumber == viewNumber) < context.M()) return;
+            if (!context.WatchOnly())
             {
-                if (!context.WatchOnly())
-                {
-                    ChangeView message = context.ChangeViewPayloads[context.MyIndex]?.GetDeserializedMessage<ChangeView>();
-                    if (message is null || message.NewViewNumber < viewNumber)
-                        localNode.Tell(new LocalNode.SendDirectly { Inventory = context.MakeChangeView(viewNumber) });
-                }
-                InitializeConsensus(viewNumber);
+                ChangeView message = context.ChangeViewPayloads[context.MyIndex]?.GetDeserializedMessage<ChangeView>();
+                if (!context.CommitSent() && (message is null || message.NewViewNumber < viewNumber || !message.Locked))
```

Comment:
Maybe include ResponseSent and something for Primary.

Reply:
Fixed.